### PR TITLE
Decouple runs_dir from cwd

### PR DESCRIPTION
**Pull request by v1kko**
- [x] create
- [x] submit
- [x] status
- [x] cleanup 
- [x] tested on the gateway
  - but gives errors, related to IDS, probably unrelated to this merge request, files look good

The most important (with relation to compatibility) is that workspace-root is not used anymore.

- create uses `create - jruns` and `create - runs_dir` to determine the location for the runs.
  - both are optional, as when we move to another system, jruns will (probably) not  be used anymore
  - a runs_dir is created by the specified system.
  - if no runs_dir is specified an incremental one is generated by create (to avoid loss of data)
  - if no `jruns` is specified, it either uses the location specified by the system (via `JRUNS`) or just the current directory
- All the other commands that need this path use the path to the runs set in the `runs.yaml`

### CODE REVIEW

**Comment by v1kko** [comment]
The most important (with relation to compatibility) is that workspace-root is not used anymore.

- create uses `create - jruns` and `create - runs_dir` to determine the location for the runs.
  - both are optional, as when we move to another system, jruns will (probably) not  be used anymore
  - a runs_dir is created by the specified system.
  - if no runs_dir is specified an incremental one is generated by create (to avoid loss of data)
  - if no `jruns` is specified, it either uses the location specified by the system (via `JRUNS`) or just the current directory
- All the other commands that need this path use the path to the runs set in the `runs.yaml`

File: src/duqtools/submit.py [hunk 1]
```diff
@@ -140,12 +140,19 @@ def submit(*, force: bool, max_jobs: int, schedule: bool, array: bool,
...
         force = True
-        jobs = [Job(Path(run)) for run in resubmit]
+        for run in resubmit:
+            if run in run_dict:
+                jobs.append(Job(run_dict[run].dirname))
+            else:
+                jobs.append(Job(Path(run)))
+
     else:
         jobs = [Job(run.dirname) for run in runs]
```
Comment (stefsmeets) [inline src/duqtools/submit.py:157]: This whole section is a bit of a mess to me. It seems like it logically mixes two workflows, I'm trying to make sense of what is happening here.

The first part tries to handle resubmissions, the second part the 'normal' submit workflow.

How about something like this:

```python
    if resubmit:
        force = True
        
        run_dict = {run.shortname: run.dirname for run in WorkDirectory().runs}

        jobs = [Job(run_dict.get(run, Path(run)) for run in resubmit]
    else:
        jobs = [Job(run.dirname) for run in WorkDirectory().runs]
```

File: src/duqtools/system.py [hunk 2]
```diff
@@ -16,7 +18,11 @@ class DummySystem(AbstractSystem):
     """
 
     @staticmethod
-    def write_batchfile(workspace: WorkDirectory, run_name: str,
+    def get_runs_dir() -> Path:
+        return Path('./')
```
Comment (stefsmeets) [inline src/duqtools/system.py:22]:
```suggestion
        return Path()
```

**Review by stefsmeets — CHANGES_REQUESTED** [review]
Nice work! If I understand correctly, with this you can create your `duqtools.yaml` anywhere, and it will automatically write to some directory defined by `$JRUNS`. If not specified it takes `create.jruns` / `create.runs_dir`. 

I'm just wondering whether these should be under `create` and not under 'system' instead in the config.

Can we make sure to copy the `duqtools.yaml` to the experiment directory? This way everything is contained.

I have left a few comments in the code with suggestions for improvements.

File: src/duqtools/jetto/_llcmd.py [hunk 3]
```diff
@@ -34,17 +30,17 @@ def write_batchfile(workspace: WorkDirectory, run_name: str,
 
     with open(llcmd_path, 'w') as f:
         f.write(f"""#!/bin/sh
-#SBATCH -J duqtools.jetto.{run_name}
+#SBATCH -J duqtools.jetto.{run_dir.name}
 #SBATCH -i /dev/null
-#SBATCH -o {full_path}/ll.out
-#SBATCH -e {full_path}/ll.err
+#SBATCH -o {run_dir}/ll.out
+#SBATCH -e {run_dir}/ll.err
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/jetto/_llcmd.py]: more code less strings!

File: src/duqtools/jetto/_system.py [hunk 4]
```diff
@@ -32,13 +32,28 @@
 class JettoSystem(AbstractSystem):
 
     @staticmethod
-    @add_to_op_queue('Writing new batchfile', '{run_name}', quiet=True)
-    def write_batchfile(workspace: WorkDirectory, run_name: str,
-                        template_drc: Path):
+    def get_runs_dir() -> Path:
+        path = WorkDirectory.jruns_path()
+        if cfg.create:  # Make mypy happy, create is Optional[]
+            runs_dir = cfg.create.runs_dir
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/jetto/_system.py:38]: strangely enough, no 🥲

File: src/duqtools/schema/cli.py [hunk 5]
```diff
@@ -71,6 +70,29 @@ class CreateConfigModel(BaseModel):
         starting by the run number given by `run_out_start_at`.
         """))
 
+    jruns: Optional[DirectoryPath] = Field(description=f(
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/schema/cli.py:73]: fixed

File: src/duqtools/cli.py [hunk 6]
```diff
@@ -182,7 +182,7 @@ def cli_create(**kwargs):
 
 
 @cli.command('recreate')
-@click.argument('runs', nargs=-1)
+@click.argument('runs', type=Path, nargs=-1)
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/cli.py:185]: fixed

File: src/duqtools/jetto/_system.py [hunk 7]
```diff
@@ -32,13 +32,28 @@
 class JettoSystem(AbstractSystem):
 
     @staticmethod
-    @add_to_op_queue('Writing new batchfile', '{run_name}', quiet=True)
-    def write_batchfile(workspace: WorkDirectory, run_name: str,
-                        template_drc: Path):
+    def get_runs_dir() -> Path:
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/jetto/_system.py:35]: fixed

File: src/duqtools/jetto/_system.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/jetto/_system.py]: removed silly code, told mypy to ignore line

File: src/duqtools/submit.py (hunk 1 above)
Comment (v1kko) [reply src/duqtools/submit.py]: created helper function, should be very clear now

File: src/duqtools/models/_workdir.py [hunk 8]
```diff
@@ -45,3 +28,19 @@ def runs(self) -> List[Run]:
...
+            raise OSError('create key not present in duqtools.yaml')
+
+        if create.jruns:
+            path: Path = create.jruns
+        else:
+            jruns = getenv('JRUNS')
+            if jruns:
+                path = Path(jruns)
+            else:
+                path = Path('./')
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/models/_workdir.py]: I dont think using the full path here is necessary, suppose its the current directory (./), then if you store that the whole directory is movable, but if you store the full path it is not movable anymore

File: src/duqtools/schema/runs.py [hunk 9]
```diff
@@ -12,11 +12,19 @@
...
     data_out: ImasBaseModel
     operations: List[Union[IDSOperation, JettoOperation,
                            List[Union[IDSOperation, JettoOperation]]]]
 
+    @root_validator()
+    def shortname_compat(cls, root):
+        # Compatibility with old runs.yaml
+        if not root['shortname']:
+            root['shortname'] = root['dirname'].name
+        return root
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/schema/runs.py:26]: after a while we should lose the compatibility validator, and I thought that we do not want to have properties in the schema

File: src/duqtools/cleanup.py [hunk 10]
```diff
@@ -42,7 +41,7 @@ def cleanup(out, force, **kwargs):
         Remove output IDS.
     """
     try:
-        workspace = WorkDirectory.parse_obj(cfg.workspace)
+        workspace = WorkDirectory()
         runs = workspace.runs
     except OSError:
         runs = ()
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/cleanup.py:47]: good catch!

File: src/duqtools/models/_workdir.py [hunk 11]
```diff
@@ -1,40 +1,23 @@
...
-
-    @property
-    def subdir(self):
-        """Get subdirectory relative to root."""
-        return self.cwd.relative_to(self.root)
-
-    @validator('root')
-    def resolve_root(cls, v):
-        return v.resolve()
+class WorkDirectory(BaseModel):
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/models/_workdir.py:10]: I don't think we have to store the `cwd`, but I do agree the WorkDirectory can be renamed to Locations, with static functions that return locations.

**Comment by v1kko** [comment]
> Just had a test and it seems to work as advertised sweat_smile
> 
> If I specify `create.runs_dir` it puts the data in `create.jruns` / `create.runs_dir` or `$JRUNS` / `create.runs_dir`.
> 
> If I do not specify `create.runs_dir`, it puts the data in `create.jruns` / `duqtools_experiment_xxxx`.
> 
> What is a bit tricky to me is, what if I want to put the data in the same directory as my `duqtools.yaml` (as we had it before). This is now a bit more difficult.
> 
> Let's say I go to $JRUNS and create a directory there with `duqtools.yaml` as before and I don't specify `runs_dir`. It now puts my data somewhere else.
> 
> What is the use-case that we want to support here?

The thought is that data is not overwritten by default, especially if you do not have direct sight of it in the JRUNS directory. and that in this way multiple duqtools runs (in parallel) are supported in the JRUNS directory.

I agree that it makes a bit less sense if the path to the runs is a subdirectory of the current working directory, I can make that behave like the old days.

**Comment by v1kko** [comment]
> Nice work! If I understand correctly, with this you can create your `duqtools.yaml` anywhere, and it will automatically write to some directory defined by `$JRUNS`. If not specified it takes `create.jruns` / `create.runs_dir`.
> 
> I'm just wondering whether these should be under `create` and not under 'system' instead in the config.
> 
> Can we make sure to copy the `duqtools.yaml` to the experiment directory? This way everything is contained.
> 
> I have left a few comments in the code with suggestions for improvements.

Storing all the files in the run directory as well, including the duqtools.yaml, but also the data.csv, runs.yaml, etc ... is a very good idea

**Comment by v1kko** [comment]
@stefsmeets  I incorporated almost all the feedback, (except where not possible).
Main change: duqtools_experiments is not created if runs_dir is equal (or a parent of) the current working directory.

File: src/duqtools/models/_locations.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+        if not runs_yaml.exists():
+            raise OSError(f'Cannot find {runs_yaml}.')
+
+        return Runs.parse_file(runs_yaml)
+
+    @property
+    def jruns_path(self) -> Path:
+        """return the Path specified in the create->jpath config variable, or,
+        if that is empty, the JPATH environment variable, or, if JPATH does not
```
Comment (stefsmeets) [inline src/duqtools/models/_locations.py:34]:
```suggestion
        """return the Path specified in the create->jruns config variable, or,
        if that is empty, the `$JRUNS` environment variable, or, if `$JRUNS` does not
```

File: src/duqtools/schema/cli.py [hunk 13]
```diff
@@ -71,6 +70,31 @@ class CreateConfigModel(BaseModel):
...
+        """Jruns defines the the root directory where all simulations are
+    run for the jetto system. Because the jettos system works with relative
+    directories from some root directory.
+
+    If this variable is not specified, duqtools will look for the `JPATH` environment
```
Comment (stefsmeets) [inline src/duqtools/schema/cli.py:78]:
```suggestion
    If this variable is not specified, duqtools will look for the `$JRUNS` environment
```

**Review by stefsmeets — APPROVED** [review]
Looks good! Let's get this merged! 🚀

**Pull request merged**